From 4138417d8015e536a2e7543fb4bb81887079d008 Mon Sep 17 00:00:00 2001
From: Jagpal Singh Gill <paligill@gmail.com>
Date: Wed, 17 Sep 2025 11:25:38 -0700
Subject: [PATCH 12/18] leakdetector: use new aserver constructor
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Use new aserver constructor as per [1] to initialize the leak detector
properties. Also, call emit_added() for the interface so interface
becomes visible on Dbus.

[1] https://github.com/openbmc/sdbusplus/commit/a0fe02cad6645d0cd9328db414758d3f2f921663

Tested:
Without Change -
```
> busctl call xyz.openbmc_project.ObjectMapper /xyz/openbmc_project/object_mapper xyz.openbmc_project.ObjectMapper GetSubTree sias "/xyz/openbmc_project/state/leak" 0 1 xyz.openbmc_project.State.Leak.Detector --json=pretty | jq .
{
  "type": "a{sa{sas}}",
  "data": [
    {}
  ]
}
```
With Change -
```
> busctl call xyz.openbmc_project.ObjectMapper /xyz/openbmc_project/object_mapper xyz.openbmc_project.ObjectMapper GetSubTree sias "/xyz/openbmc_project/state/leak" 0 1 xyz.openbmc_project.State.Leak.Detector --json=pretty | jq .
{
  "type": "a{sa{sas}}",
  "data": [
    {
      "/xyz/openbmc_project/state/leak/detector/ComputeBlade": {
        "xyz.openbmc_project.leakdetector": [
          "xyz.openbmc_project.State.Leak.Detector"
        ]
      }
    }
  ]
}

> busctl introspect xyz.openbmc_project.leakdetector /xyz/openbmc_project/state/leak/detector/ComputeBlade
NAME                                        TYPE      SIGNATURE RESULT/VALUE                             FLAGS
org.freedesktop.DBus.Introspectable         interface -         -                                        -
.Introspect                                 method    -         s                                        -
org.freedesktop.DBus.Peer                   interface -         -                                        -
.GetMachineId                               method    -         s                                        -
.Ping                                       method    -         -                                        -
org.freedesktop.DBus.Properties             interface -         -                                        -
.Get                                        method    ss        v                                        -
.GetAll                                     method    s         a{sv}                                    -
.Set                                        method    ssv       -                                        -
.PropertiesChanged                          signal    sa{sv}as  -                                        -
xyz.openbmc_project.Association.Definitions interface -         -                                        -
.Associations                               property  a(sss)    0                                        emits-change writable
xyz.openbmc_project.State.Leak.Detector     interface -         -                                        -
.PrettyName                                 property  s         "ComputeBlade"                           emits-change
.State                                      property  s         "xyz.openbmc_project.State.Leak.Detecto… emits-change
.Type                                       property  s         "xyz.openbmc_project.State.Leak.Detecto… emits-change writable
```

Change-Id: I1dc982f4621301143e48380eb4168298dd63efdf
Signed-off-by: Jagpal Singh Gill <paligill@gmail.com>
---
 src/leakdetector/LeakGPIODetector.cpp | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/src/leakdetector/LeakGPIODetector.cpp b/src/leakdetector/LeakGPIODetector.cpp
index aedffa1..fe4d20d 100644
--- a/src/leakdetector/LeakGPIODetector.cpp
+++ b/src/leakdetector/LeakGPIODetector.cpp
@@ -5,6 +5,8 @@
 
 #include <sdbusplus/async.hpp>
 #include <sdbusplus/message/native_types.hpp>
+#include <xyz/openbmc_project/Association/Definitions/aserver.hpp>
+#include <xyz/openbmc_project/State/Leak/Detector/aserver.hpp>
 
 #include <array>
 #include <functional>
@@ -42,14 +44,16 @@ static auto getObjectPath(const std::string& detectorName)
 
 GPIODetector::GPIODetector(sdbusplus::async::context& ctx, Events& leakEvents,
                            const config::DetectorConfig& config) :
-    DetectorIntf(ctx, getObjectPath(config.name).str.c_str()), ctx(ctx),
-    leakEvents(leakEvents), config(config),
+    DetectorIntf(ctx, getObjectPath(config.name).str.c_str(),
+                 DetectorIntf::Definitions::properties_t{},
+                 DetectorIntf::Detector::properties_t{
+                     config.name, DetectorState::Normal, config.type}),
+    ctx(ctx), leakEvents(leakEvents), config(config),
     gpioInterface(ctx, config.name, config.pinName,
                   (config.polarity == config::PinPolarity::activeLow),
                   std::bind_front(&GPIODetector::updateGPIOStateAsync, this))
 {
-    pretty_name<false>(config.name);
-    type<false>(config.type);
+    Detector::emit_added();
 
     ctx.spawn(gpioInterface.start());
 
-- 
2.47.3

